assignment-client/src/assets/AssetServer.cpp

```diff
@@ -0,0 +1,278 @@
+}
+
+void AssetServer::run() {
+    ThreadedAssignment::commonInit(ASSET_SERVER_LOGGING_TARGET_NAME, NodeType::AssetServer);
+
+    auto nodeList = DependencyManager::get<NodeList>();
+    nodeList->addNodeTypeToInterestSet(NodeType::Agent);
+
+    _resourcesDirectory = QDir(QCoreApplication::applicationDirPath()).filePath("resources/assets");
+    if (!_resourcesDirectory.exists()) {
+        qDebug() << "Creating resources directory";
+        _resourcesDirectory.mkpath(".");
+    }
+    qDebug() << "Serving files from: " << _resourcesDirectory.path();
+
+    // Scan for new files
+    qDebug() << "Looking for new files in asset directory";
+    auto files = _resourcesDirectory.entryInfoList(QDir::Files);
+    QRegExp filenameRegex { "^[a-f0-9]{32}(\\..+)?$" };
+    for (auto fileInfo : files) {
+        auto filename = fileInfo.fileName();
+        if (!filenameRegex.exactMatch(filename)) {
+            qDebug() << "Found file: " << filename;
+            if (!fileInfo.isReadable()) {
+                qDebug() << "\tCan't open file for reading: " << filename;
+                continue;
+            }
+
+            // Read file
+            QFile file { fileInfo.absoluteFilePath() };
+            file.open(QFile::ReadOnly);
+            QByteArray data = file.readAll();
+
+            auto hash = hashData(data);
+
+            qDebug() << "\tMoving " << filename << " to " << hash;
+
+            file.rename(_resourcesDirectory.absoluteFilePath(hash));
+        }
+    }
+
+    while (!_isFinished) {
+        // since we're a while loop we need to help Qt's event processing
+        QCoreApplication::processEvents();
+    }
+}
+
+void AssetServer::handleAssetGetInfo(QSharedPointer<NLPacket> packet, SharedNodePointer senderNode) {
+    if (packet->getPayloadSize() < HASH_HEX_LENGTH) {
+        qDebug() << "ERROR bad file request";
+        return;
+    }
+
+    QByteArray assetHash;
+    MessageID messageID;
+    packet->readPrimitive(&messageID);
+    assetHash = packet->read(HASH_HEX_LENGTH);
+
+    auto replyPacket = NLPacket::create(PacketType::AssetGetInfoReply);
+
+    replyPacket->writePrimitive(messageID);
+    replyPacket->write(assetHash, HASH_HEX_LENGTH);
+
+    QFileInfo fileInfo { _resourcesDirectory.filePath(QString(assetHash)) };
+    qDebug() << "Opening file: " << QString(QFileInfo(assetHash).fileName());
+
+    if (fileInfo.exists() && fileInfo.isReadable()) {
+        replyPacket->writePrimitive(AssetServerError::NO_ERROR);
+        replyPacket->writePrimitive(fileInfo.size());
+    } else {
+        qDebug() << "Asset not found: " << assetHash;
+        replyPacket->writePrimitive(AssetServerError::ASSET_NOT_FOUND);
+    }
+
+    auto nodeList = DependencyManager::get<NodeList>();
+    nodeList->sendPacket(std::move(replyPacket), *senderNode);
+}
+
+void AssetServer::handleAssetGet(QSharedPointer<NLPacket> packet, SharedNodePointer senderNode) {
+    if (packet->getPayloadSize() < HASH_HEX_LENGTH) {
```

Comment:
I think we can improve error handling in general. I have some thoughts on it that we can discuss for a future PR.